### PR TITLE
Remove undefined behavior for address of null reference

### LINKED ISSUE #73088: MethodSignature undefined use of  reference binding to null pointer type (per gcc -fsanitize=undefined)
### Description

The constructors for MethodSignature allow for member m_pSubst to be assigned NULL.

Later, MethodSignature::GetSubstition is called, which returns a const reference to the contents of m_pSubst.  You can't take a reference of a NULL object, as that is undefined behavior.

It probably all works out, through the maze of * and &.

### Reproduction Steps

Run gcc -fsanitize=undefined

### Expected behavior

This is undefined behavior.

### Actual behavior

It probably all works out for this code.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Fixes #73088

### CODE REVIEW

File: src/coreclr/vm/methodtablebuilder.h [hunk 1]
```diff
@@ -783,9 +783,9 @@ class MethodTableBuilder
 
         //-----------------------------------------------------------------------------------------
         // Returns the substitution to be used in interpreting the signature.
-        const Substitution &
+        const Substitution *
         GetSubstitution() const
```
Comment (am11) [inline src/coreclr/vm/methodtablebuilder.h:787]: Should we make the same change in `src/coreclr/vm/methodtablebuilder.inl`?

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 2]
```diff
@@ -6599,7 +6599,7 @@ VOID MethodTableBuilder::PlaceInterfaceDeclarationOnClass(
     DispatchMapTypeID firstDispatchMapTypeID;
     ComputeDispatchMapTypeIDs(
         pDeclMT,
-        &pDecl->GetMethodSignature().GetSubstitution(),
+        pDecl->GetMethodSignature().GetSubstitution(),
```
Comment (am11) [inline src/coreclr/vm/methodtablebuilder.cpp:6602]: Looks like there are a few other instances of this function call under `#if DEBUG`, which needs this fix `git grep -n 'GetSubstitution('`

Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/methodtablebuilder.cpp:6602]: Boo. I thought I was building under debug locally. :(

File: src/coreclr/vm/methodtablebuilder.h (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/vm/methodtablebuilder.h:787]: Nope. That is for a different type and seems to be "correct". I'd prefer to leave this PR as simply making things correct and then we can iterate on improving the contracts all up.

**Review by jkotas — APPROVED** [review]

**Pull request merged**